state/executor.go

```diff
@@ -0,0 +1,229 @@
+			IntermediateStateRoots: types.IntermediateStateRoots{RawRootsList: nil},
+			Evidence:               types.EvidenceData{Evidence: nil},
+		},
+	}
+
+	return block
+}
+
+func (e *BlockExecutor) ApplyBlock(state State, block *types.Block) (State, uint64, error) {
+	err := e.validate(state, block)
+	if err != nil {
+		return State{}, 0, err
+	}
+
+	resp, err := e.execute(state, block)
+	if err != nil {
+		return State{}, 0, err
+	}
+
+	state, err = e.updateState(state, block, resp)
+	if err != nil {
+		return State{}, 0, err
+	}
+
+	appHash, retainHeight, err := e.commit(state, block, resp.DeliverTxs)
+	if err != nil {
+		return State{}, 0, err
+	}
+
+	copy(state.AppHash[:], appHash[:])
+
+	return state, retainHeight, nil
+}
+
+func (e *BlockExecutor) updateState(state State, block *types.Block, abciResponses *tmstate.ABCIResponses) (State, error) {
+	// TODO(tzdybal): blockID
+	s := State{
+		Version:         state.Version,
+		ChainID:         state.ChainID,
+		InitialHeight:   state.InitialHeight,
+		LastBlockHeight: int64(block.Header.Height),
+		LastBlockTime:   time.Unix(int64(block.Header.Time), 0),
+		// skipped all "Validators" fields
+		ConsensusParams:                  state.ConsensusParams,
+		LastHeightConsensusParamsChanged: state.LastHeightConsensusParamsChanged,
+	}
+	copy(s.LastResultsHash[:], lltypes.NewResults(abciResponses.DeliverTxs).Hash())
+
+	return s, nil
+}
+
+func (e *BlockExecutor) commit(state State, block *types.Block, deliverTxs []*abci.ResponseDeliverTx) ([]byte, uint64, error) {
+	e.mempool.Lock()
+	defer e.mempool.Unlock()
+
+	err := e.mempool.FlushAppConn()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	resp, err := e.proxyApp.CommitSync(context.TODO())
+	if err != nil {
+		return nil, 0, err
+	}
+
+	maxBytes := state.ConsensusParams.Block.MaxBytes
+	maxGas := state.ConsensusParams.Block.MaxGas
+	err = e.mempool.Update(int64(block.Header.Height), fromOptimintTxs(block.Data.Txs), deliverTxs, mempool.PreCheckMaxBytes(maxBytes), mempool.PostCheckMaxGas(maxGas))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return resp.Data, uint64(resp.RetainHeight), err
+}
+
+func (e *BlockExecutor) validate(state State, block *types.Block) error {
+	if block.Header.Version.App != state.Version.Consensus.App ||
+		block.Header.Version.Block != state.Version.Consensus.Block {
+		return errors.New("block version mismatch")
+	}
+	if state.LastBlockHeight <= 0 && block.Header.Height != uint64(state.InitialHeight) {
+		return errors.New("initial block height mismatch")
+	}
+	if state.LastBlockHeight > 0 && block.Header.Height != uint64(state.LastBlockHeight)+1 {
+		return errors.New("block height mismatch")
+	}
+	if !bytes.Equal(block.Header.AppHash[:], state.AppHash[:]) {
+		return errors.New("AppHash mismatch")
+	}
+
+	if !bytes.Equal(block.Header.LastResultsHash[:], state.LastResultsHash[:]) {
+		return errors.New("LastResultsHash mismatch")
+	}
+
+	// TODO(tzdybal): what else to check
```

Comment:
perhaps we could have a `ValidateBasic` method for `types.Block` and call that here too. ref https://github.com/lazyledger/lazyledger-core/blob/8da16447f34680639033b531ed10b9ab756a14e0/types/block.go#L168-L168